state/protocol/badger/state.go

```diff
@@ -167,55 +141,67 @@ func Bootstrap(
+			}
+
+			// initialize version beacon
+			err = boostrapVersionBeacon(rw, root)
+			if err != nil {
+				return fmt.Errorf("could not bootstrap version beacon: %w", err)
+			}
+
+			return nil
+		})
 		if err != nil {
-			return fmt.Errorf("could not bootstrap version beacon: %w", err)
+			return fmt.Errorf("bootstrapping failed: %w", err)
 		}
 
-		return nil
-	})
-	if err != nil {
-		return nil, fmt.Errorf("bootstrapping failed: %w", err)
-	}
+		// CAUTION: INSERT FINALIZED HEIGHT must be LAST, because we use its existence in the database
+		// as indicator that the protocol database has been bootstrapped successfully. Before we write the
+		// final piece of data to complete the bootstrapping, we query the current state of the database
+		// (sanity check) to ensure that it is still considered as not properly bootstrapped.
+		isBootstrapped, err = IsBootstrapped(db)
+		if err != nil {
+			return fmt.Errorf("determining whether database is successfully bootstrapped failed with unexpected exception: %w", err)
+		}
+		if isBootstrapped { // we haven't written the latest finalized height yet, so this value must be false
+			return fmt.Errorf("sanity check failed: while bootstrapping has not yet completed, the implementation already considers the protocol state as successfully bootstrapped")
+		}
+		err = db.WithReaderBatchWriter(func(rw storage.ReaderBatchWriter) error {
+			// initialize the current protocol state height/view pointers
+			return bootstrapStatePointers(lctx, rw, root)
+		})
+		if err != nil {
+			return fmt.Errorf("could not bootstrap height/view pointers: %w", err)
+		}
 
-	// CAUTION: INSERT FINALIZED HEIGHT must be LAST, because we use its existence in the database
-	// as indicator that the protocol database has been bootstrapped successfully. Before we write the
-	// final piece of data to complete the bootstrapping, we query the current state of the database
-	// (sanity check) to ensure that it is still considered as not properly bootstrapped.
-	isBootstrapped, err = IsBootstrapped(db)
-	if err != nil {
-		return nil, fmt.Errorf("determining whether database is successfully bootstrapped failed with unexpected exception: %w", err)
-	}
-	if isBootstrapped { // we haven't written the latest finalized height yet, so this vaule must be false
-		return nil, fmt.Errorf("sanity check failed: while bootstrapping has not yet completed, the implementation already considers the protocol state as successfully bootstrapped")
-	}
-	err = db.WithReaderBatchWriter(func(rw storage.ReaderBatchWriter) error {
-		// initialize the current protocol state height/view pointers
-		return bootstrapStatePointers(lctx, rw, root)
+		state, err = OpenState(metrics, db, lockManager, headers, seals, results, blocks, qcs, setups, commits, epochProtocolStateSnapshots, protocolKVStoreSnapshots, versionBeacons)
+		if err != nil {
+			return fmt.Errorf("bootstrapping failed, because the resulting database state is rejected: %w", err)
+		}
```

Comment:
From a causal perspective, I think it would improve clarity if we moved the `OpenState` call _after_ the [locking scope](https://github.com/onflow/flow-go/blob/50b4ba44c396fc4823f018b640b7ea21d9b304be/state/protocol/badger/state.go#L149-L205) required to _initialize_ the state. Conceptually, only a properly initialized state (after which the locks should be released) should be opened.